### ISSUE TITLE
job control enhancements

### ISSUE THREAD

**Issue by justinmk** [neovim/neovim#1869]
Some enhancements for job control:

- return exit status. https://github.com/neovim/neovim/pull/1844
- migrate `job_data` to a dict instead of a list
- rename `job_data` to `job` ("data" is almost always redundant on a variable name, and in this case it's particularly important to keep the name brief, as well as avoid underscores, because users are encouraged to use vimscript for ad-hoc one-liner commands)
- jobstop() and jobsend() functions should accept the job name instead of job id. ([ref](https://github.com/junegunn/vim-plug/pull/103#issuecomment-58566485))
- add a `jobwait(jobids...)` function to wait for multiple jobs and return a list of exit status codes ([ref](https://github.com/junegunn/vim-plug/pull/103#issuecomment-58603245))

**Comment by tarruda** [comment]
> add a jobwait(jobids...) function to wait for multiple jobs and return a list of exit status codes

Another option is to return the status code as the last array element of the "exit" event, the user can always append that to a list if require.

Here's something I've been thinking about.

vimscript has an API that is driven by global state(a consequence of how the code itself is organized) and IMO autocommands are a consequence of that(no lexical scope, arguments are passed as global variables and/or substituted strings such as `<args>`)

The initial job API was designed to be as idiomatic as possible to the classic vimscript programmer, which makes heavy use of autocommands, but maybe it wouldn't be such a bad idea if we diverge from that by using function callbacks.

For example, here's my [job control demo](https://gist.github.com/tarruda/1b5010531bb6ae7382a4) adapted to a callback-based API:

```vim
 function s:JobEvent(job_id, event_id, event_arg)
  if a:event_id == 'stdout'
    let str = 'Message from job '.a:job_id.': '.a:event_arg
  elseif a:event_id == 'stderr'
    let str = 'Error message from job '.a:job_id.': '.a:event_arg
  else
    let str = 'Job '.a:job_id.' exited with status '.a:event_arg
  endif
 
  call append(line('$'), str)
endfunction

:let srv1_id = jobstart('nc', ['-l', '9991'], funcref('s:JobEvent'))
:let srv2_id = jobstart('nc', ['-l', '9991'], funcref('s:JobEvent'))
```

This results in a cleaner API:

- JobEvent autocommand removed
- v:job_data variable removed
- There's no need to have a job name to match as autocommand pattern.

What do you think?

**Comment by Grimy** [comment]
While the

**Comment by Grimy** [comment]
@tarruda How would you remove callbacks with your proposed API? With the current API it’s as simple as ```au! JobActivity pattern```.

**Comment by tarruda** [comment]
> @tarruda How would you remove callbacks with your proposed API? With the current API it’s as simple as au! JobActivity pattern.

There's no way to detach, but that can easily be handled in the callback itself: if the job reached a state where the callback should be detached, the user can simply return early from the function.

**Comment by tarruda** [comment]
This can be closed with the refactor in #2247,
